src/dmd/tocsym.d

```diff
@@ -201,6 +201,29 @@ Symbol *toSymbol(Dsymbol s)
                 t.Tcount++;
             }
 
+            /* Even if a symbol is immutable, if it has a constructor then
+             * the constructor mutates it. Remember that constructors can
+             * be inlined into other code.
+             * Just can't rely on it being immutable.
+             */
+            if (t.Tty & (mTYimmutable | mTYconst))
+            {
+                if (auto ts = vd.type.isTypeStruct())
+                {
+                    if (!ts.isMutable() && ts.sym.ctor)
+                    {
+                        t = type_setty(&t, t.Tty & ~(mTYimmutable | mTYconst));
+                    }
+                }
+                else if (auto tc = vd.type.isTypeClass())
+                {
+                    if (!tc.isMutable() && tc.sym.ctor)
+                    {
+                        t = type_setty(&t, t.Tty & ~(mTYimmutable | mTYconst));
+                    }
+                }
+            }
+
```

Comment:
AFAICS, the code above aims to address the following cases:
```D
struct S
{
    int v = 1;
    this(int x) { v = x * 2; }
}

struct C
{
    int v = 1;
    this(int x) { v = x * 2; }
}

void test()
{
    // 1)
    const S cs = S(7);  // cs.v set to 1 before constructor call
    assert(cs.v == 14); // cs.v could be 1 because of CSE

    // 2)
    const C cc = C(7);  // cc.v set to 1 before constructor call
    assert(cc.v == 14); // cc.v could be 1 because of CSE
}
```

However I don't see the following cases addressed:

```D
// 3)
immutable int x;
shared static this()
{
    x += 3;
}

// 4)
struct S
{
    immutable int v;
    this(int x) { v += v; }
}
```

Reply:
You're right. Will work on a fix.

Reply:
@ZombineDev resolved.

Reply:
Thanks!